### PR TITLE
[IOTDB-165][TsFile] Delete a current version and add a number version and an exception.

### PR DESCRIPTION
**Pull request by Genius-pig**
According to this [issue](https://issues.apache.org/jira/projects/IOTDB/issues/IOTDB-165), I implement 1, 2, delete a current version which I think is useless, and I will add an exception for TsFile reader in a few days. But discussion of the compatibility between iotdb and TsFile is still continuing.

> 1. Change current TsFile magic number from 12 bytes to: 6 bytes magic string ("TsFile") + 6 bytes version number ({"000001", "000002", ""000003"}) ("v0.8.0" is the first version (we treat it as "000000")).We update the TsFile version unless the format is incompatible.

>2. The tail of a TsFile only has "TsFile" magic string, without the version.

### CODE REVIEW

**Comment by Genius-pig** [comment]
It failed in windows platform, it says can't find file. Why did this happen?

**Comment by Genius-pig** [comment]
It failed in windows platform, there were some error informations, but I don't know why.
```sh
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.014 s <<< FAILURE! - in org.apache.iotdb.tsfile.read.TsFileSequenceReaderTest
[ERROR] testReadTsFileSequently(org.apache.iotdb.tsfile.read.TsFileSequenceReaderTest)  Time elapsed: 0.014 s  <<< ERROR!
java.io.FileNotFoundException: target\testTsFile.tsfile (Access is denied)
	at org.apache.iotdb.tsfile.read.TsFileSequenceReaderTest.before(TsFileSequenceReaderTest.java:54)
[ERROR] testReadTsFileSequently(org.apache.iotdb.tsfile.read.TsFileSequenceReaderTest)  Time elapsed: 0.014 s  <<< ERROR!
java.lang.NullPointerException
	at org.apache.iotdb.tsfile.read.TsFileSequenceReaderTest.after(TsFileSequenceReaderTest.java:61)
```

**Comment by Genius-pig** [comment]
@qiaojialin @jixuan1989 please take a look.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/common/conf/TSFileConfig.java [hunk 1]
```diff
@@ -1,187 +1,186 @@
...
+package org.apache.iotdb.tsfile.common.conf;
+
+import org.apache.iotdb.tsfile.fileSystem.FSType;
+
+/**
+ * TSFileConfig is a configure class. Every variables is public and has default value.
+ *
+ * @author kangrong
+ */
+public class TSFileConfig {
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/common/conf/TSFileConfig.java]: why this file is changed totally...

Comment (Genius-pig) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/common/conf/TSFileConfig.java]: @jixuan1989 Already fixed it. I merge all commits into one commit.

**Comment by Genius-pig** [comment]
@jixuan1989 I think don't need it, because when reading the head magic, you have already known the compatibility. The tail magic is used for an end sign. BTW, it's just my opinions.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TsFileMetaData.java [hunk 2]
```diff
@@ -112,8 +105,6 @@ public static TsFileMetaData deserializeFrom(InputStream inputStream) throws IOE
       }
     }
 
-    fileMetaData.currentVersion = ReadWriteIOUtils.readInt(inputStream);
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TsFileMetaData.java]: Remind to maintain the format-changelist.md

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 3]
```diff
@@ -195,8 +196,8 @@ public String readTailMagic() throws IOException {
    * whether the file is a complete TsFile: only if the head magic and tail magic string exists.
    */
   public boolean isComplete() throws IOException {
-    return tsFileInput.size() >= TSFileConfig.MAGIC_STRING.length() * 2 && readTailMagic()
-        .equals(readHeadMagic());
+    return tsFileInput.size() >= TSFileConfig.MAGIC_STRING.length() * 2 + TSFileConfig.VERSION_NUMBER.length() &&
+        (readTailMagic() + TSFileConfig.VERSION_NUMBER).equals(readHeadMagic());
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java]: In my opinion, I think we can use two methods: `readHeadMagic()` which return "TsFile" and `readVersionNumber()`. We do not need to do operations like "String + String" (e.g., `readTailMagic() + TSFileConfig.VERSION_NUMBER`), which waste memory..

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 4]
```diff
@@ -213,15 +214,26 @@ public String readHeadMagic() throws IOException {
    * to the end of the magic head string.
    */
   public String readHeadMagic(boolean movePosition) throws IOException {
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java]: throws NotCompatibleException,  IOException

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 5]
```diff
@@ -213,15 +214,26 @@ public String readHeadMagic() throws IOException {
...
-    return new String(magicStringBytes.array());
+    String headMagic = new String(magicStringBytes.array());
+    checkHeadMagic(headMagic);
+    return headMagic;
+  }
+
+  /**
+   * this function checks compatibility of TsFile.
+   */
+  public void checkHeadMagic(String headMagic) throws IOException {
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java]: throws NotCompatibleException

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 6]
```diff
@@ -213,15 +214,26 @@ public String readHeadMagic() throws IOException {
...
+    if(!headMagic.equals(TSFileConfig.MAGIC_STRING + TSFileConfig.VERSION_NUMBER)) {
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java]: same question, do not need to generate a new String by using +.

File: tsfile/src/test/java/org/apache/iotdb/tsfile/read/filter/IExpressionOptimizerTest.java [hunk 7]
```diff
@@ -96,9 +95,8 @@ public void testSeriesOnly() {
...
           .and(BinaryExpression.or(singleSeriesExp1, singleSeriesExp2),
               singleSeriesExp3);
-      Assert.assertEquals(true,
-          expression.toString()
-              .equals(expressionOptimizer.optimize(expression, selectedSeries).toString()));
+      Assert.assertEquals(expression.toString(),
+          expressionOptimizer.optimize(expression, selectedSeries).toString());
 
     } catch (QueryFilterOptimizationException e) {
       e.printStackTrace();
```
Comment (jixuan1989) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/read/filter/IExpressionOptimizerTest.java]: Can you replace this e.print  with logger.error()?

Comment (Genius-pig) [reply tsfile/src/test/java/org/apache/iotdb/tsfile/read/filter/IExpressionOptimizerTest.java]: Use a Assert.fail() is better?

**Pull request closed without merging**